File: backend/src/main/java/com/codeloom/backend/converter/AbstractJsonWritingConverter.java
package com.codeloom.backend.converter;

import lombok.RequiredArgsConstructor;
import org.jspecify.annotations.NonNull;
import org.postgresql.util.PGobject;
import org.springframework.core.convert.converter.Converter;
import org.springframework.data.jdbc.core.mapping.JdbcValue;
import tools.jackson.databind.ObjectMapper;

import java.sql.JDBCType;
import java.sql.SQLException;

@RequiredArgsConstructor
public abstract class AbstractJsonWritingConverter<T> implements Converter<T, JdbcValue> {
    private final ObjectMapper mapper;

    @Override
    public JdbcValue convert(@NonNull T source) {
        try {
            PGobject obj = new PGobject();
            obj.setType("jsonb");
            obj.setValue(mapper.writeValueAsString(source));
            return JdbcValue.of(obj, JDBCType.OTHER);
        } catch (SQLException e) {
            throw new IllegalArgumentException(e);
        }
    }
}

File: backend/src/main/kotlin/com/codeloom/backend/converter/AbstractJsonWritingConverter.kt
package com.codeloom.backend.converter

import com.fasterxml.jackson.databind.ObjectMapper
import org.postgresql.util.PGobject
import org.springframework.core.convert.converter.Converter
import org.springframework.data.jdbc.core.mapping.JdbcValue
import java.sql.JDBCType

abstract class AbstractJsonWritingConverter<T : Any>(
    protected val objectMapper: ObjectMapper,
    protected val clazz: Class<T>,
) : Converter<T, JdbcValue> {

    override fun convert(source: T): JdbcValue {
        val obj = PGobject()
        obj.type = "jsonb"
        obj.value = objectMapper.writeValueAsString(source)
        return JdbcValue.of(obj, JDBCType.OTHER)
    }
}